src/measurement_buffer.cpp

```diff
@@ -150,53 +129,46 @@ void MeasurementBuffer::BufferPCLCloud(const \
-    pcl_ros::transformPointCloud(cloud, *cld_global, tf_transform);
-    cld_global->header.stamp = cloud.header.stamp;
+    // transform the cloud in the global frame
+    point_cloud_ptr cld_global(new sensor_msgs::PointCloud2());
+    geometry_msgs::TransformStamped tf_stamped = _buffer.lookupTransform(_global_frame, cloud.header.frame_id, cloud.header.stamp);
+    tf2::doTransform (cloud, *cld_global, tf_stamped);
+
+    pcl::PCLPointCloud2::Ptr cloud_pcl (new pcl::PCLPointCloud2 ());
+    pcl::PCLPointCloud2::Ptr cloud_filtered (new pcl::PCLPointCloud2 ());
 
-    // if user wants to use a voxel filter
+    pcl_conversions::toPCL(*cld_global, *cloud_pcl);
+
+    // remove points that are below or above our height restrictions, and in the same time, remove NaNs
+    // and if user wants to use it, combine with a voxel filter
     if ( _voxel_filter )
     {
-      // remove nans because they show things down
-      point_cloud_ptr cld_no_nan(new pcl::PointCloud<pcl::PointXYZ>);
-      std::vector<int> indices;
-      pcl::removeNaNFromPointCloud(*cld_global, *cld_no_nan, indices);
-
-      // minimize information needed to process
-      cld_global->clear();
-      pcl::ApproximateVoxelGrid<pcl::PointXYZ> sor1;
-      sor1.setInputCloud (cld_no_nan);
-      sor1.setLeafSize ((float)_voxel_size,
-                        (float)_voxel_size,
-                        (float)_voxel_size);
-      sor1.filter (*cld_global);
+        pcl::VoxelGrid<pcl::PCLPointCloud2> sor;
+        sor.setInputCloud (cloud_pcl);
+        sor.setFilterFieldName("z");
+        sor.setFilterLimits(_min_obstacle_height,_max_obstacle_height);
+        sor.setDownsampleAllData(false);
+        sor.setLeafSize ((float)_voxel_size,
+                         (float)_voxel_size,
+                         (float)_voxel_size);
+        sor.filter(*cloud_filtered);
+        cloud_filtered.swap(cloud_pcl);
+        cloud_pcl->is_dense=true;
     }
-
-    // remove points that are below or above our height restrictions
-    pcl::PointCloud<pcl::PointXYZ>& obs_cloud = \
-                                *(_observation_list.front()._cloud);
-    unsigned int cloud_size = cld_global->points.size();
-
-    obs_cloud.points.resize(cloud_size);
-    unsigned int point_count = 0;
-    pcl::PointCloud<pcl::PointXYZ>::iterator it;
-    for (it = cld_global->begin(); it != cld_global->end(); ++it)
+    else
     {
-      if (it->z <= _max_obstacle_height && it->z >= _min_obstacle_height)
-      {
-        obs_cloud.points.at(point_count++) = *it;
-      }
+        pcl::PassThrough<pcl::PCLPointCloud2> pass_through_filter;
+        pass_through_filter.setInputCloud(cloud_pcl);
+        pass_through_filter.setKeepOrganized(false);
+        pass_through_filter.setFilterFieldName("z");
+        pass_through_filter.setFilterLimits(_min_obstacle_height,_max_obstacle_height);
+        pass_through_filter.filter(*cloud_filtered);
+        cloud_filtered.swap(cloud_pcl);
```

Comment:
Is there a reason you swap pointers vs. just operating on the same pcl pointer? I dont see much benefit from having 2 pointers and one is just being copied out

Reply:
PCL filters requires to have an input and an output cloud pointer.
However I now removed the swap and used cloud_filtered as an input of the next conversion (reason for the swap was having the possibility to comment the whole filter block during testing).